connector-pagerduty/src/main/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyTransformer.java

```diff
@@ -45,30 +48,43 @@ public class PagerDutyTransformer implements Processor {
     @Override
     public void process(Exchange exchange) {
         JsonObject cloudEventPayload = exchange.getIn().getBody(JsonObject.class);
+        validatePayload(cloudEventPayload);
 
         JsonObject message = new JsonObject();
         message.put(EVENT_ACTION, PagerDutyEventAction.TRIGGER);
         message.mergeIn(getClientLink(cloudEventPayload, cloudEventPayload.getString(ENVIRONMENT_URL)));
 
         JsonObject messagePayload = new JsonObject();
         messagePayload.put(SUMMARY, cloudEventPayload.getString(EVENT_TYPE));
-        messagePayload.put(TIMESTAMP, LocalDateTime.parse(cloudEventPayload.getString(TIMESTAMP)).format(PD_DATE_TIME_FORMATTER));
+
+        String timestamp = cloudEventPayload.getString(TIMESTAMP);
+        if (timestamp != null) {
+            try {
+                messagePayload.put(TIMESTAMP, LocalDateTime.parse(timestamp).format(PD_DATE_TIME_FORMATTER));
+            } catch (DateTimeParseException e) {
+                Log.warnf(e, "Unable to parse timestamp %s, dropped from payload", timestamp);
+            } catch (DateTimeException e) {
+                Log.warnf(e, "Timestamp %s was successfully parsed, but could not be reformatted for PagerDuty, dropped from payload", timestamp);
+            }
+        } else {
+            Log.warn("Timestamp not provided for payload, ignored");
+        }
+
         messagePayload.put(SEVERITY, PagerDutySeverity.fromJson(cloudEventPayload.getString(SEVERITY)));
         messagePayload.put(SOURCE, cloudEventPayload.getString(APPLICATION));
         messagePayload.put(GROUP, cloudEventPayload.getString(BUNDLE));
 
-        JsonObject cloudSource = cloudEventPayload.getJsonObject(SOURCE);
-        JsonObject sourceNames = JsonObject.of(
-                APPLICATION, cloudSource.getJsonObject(APPLICATION).getString(DISPLAY_NAME),
-                BUNDLE, cloudSource.getJsonObject(BUNDLE).getString(DISPLAY_NAME),
-                EVENT_TYPE, cloudSource.getJsonObject(EVENT_TYPE).getString(DISPLAY_NAME)
-        );
-
         JsonObject customDetails = new JsonObject();
         customDetails.put(ACCOUNT_ID, cloudEventPayload.getString(ACCOUNT_ID));
         customDetails.put(ORG_ID, cloudEventPayload.getString(ORG_ID));
-        customDetails.put(CONTEXT, JsonObject.mapFrom(cloudEventPayload.getJsonObject(CONTEXT)));
-        customDetails.put(SOURCE_NAMES, sourceNames);
+        customDetails.put(CONTEXT, cloudEventPayload.getJsonObject(CONTEXT));
+
+        // Add source names, if provided
+        JsonObject cloudSource = getSourceNames(cloudEventPayload.getJsonObject(SOURCE));
+        if (cloudSource != null) {
+            customDetails.put(SOURCE, cloudSource);
+        }
```

Comment:
Same thing about the source. The bundle, application and event type are fields that will always be present when coming from the engine, so I would say it's safe to assume that they'll be there for the connector.

Reply:
Will update as well

Reply:
Actually I think this one has to stay, because the display names (`.payload.source.bundle.display_name`) are not a required field. For example, [test notifications don't include them](https://github.com/RedHatInsights/notifications-backend/blob/01b015725aaf585764872a9311e56a19835ca628/connector-pagerduty/src/test/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyTransformerTest.java#L89-L116).

Reply:
The actual names (`.payload.bundle`) are required, so they don't need a check ([like here](https://github.com/RedHatInsights/notifications-backend/blob/a384554c31289549afea76d9dbca9cd81061d279/connector-pagerduty/src/main/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyTransformer.java#L75)).

Reply:
turns out I accidentally changed a field name here (meant to be `source_names` in the final payload, not `source`), that might have caused some of the confusion.

Reply:
My point was that the `BaseTransformer` always adds those fields to the `source` key:

https://github.com/RedHatInsights/notifications-backend/blob/73af1df3cfdd8b7ab057a48069e0eff78099f232/engine/src/main/java/com/redhat/cloud/notifications/transformers/BaseTransformer.java#L62-L76

And that the `PagerDutyProcessor` uses that transformer too before sending the events through Kafka to the connector:

https://github.com/RedHatInsights/notifications-backend/blob/73af1df3cfdd8b7ab057a48069e0eff78099f232/engine/src/main/java/com/redhat/cloud/notifications/processors/pagerduty/PagerDutyProcessor.java#L75

That's why I think I'm missing how the connector would receive a payload without that data :thinking:

Reply:
That's what I thought, but I discovered that test notifications (sent by `POST /endpoints/{uuid}/test`) don't include those fields:

https://github.com/RedHatInsights/notifications-backend/blob/01b015725aaf585764872a9311e56a19835ca628/connector-pagerduty/src/test/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyTransformerTest.java#L92-L112

It looks like the test endpoint bypasses the `PagerDutyProcessor` completely, and just injects its own cloud event into the Kafka topic?

https://github.com/RedHatInsights/notifications-backend/blob/45eef7f93689ad3615b6a2cfa229dc98f072edc1/engine/src/main/java/com/redhat/cloud/notifications/routers/EndpointTestResource.java#L34-L56

https://github.com/RedHatInsights/notifications-backend/blob/45eef7f93689ad3615b6a2cfa229dc98f072edc1/common/src/main/java/com/redhat/cloud/notifications/models/event/TestEventHelper.java#L60-L93

That would explain why the fields are missing...